File: modelonly.js
// model.js
const tf = require('@tensorflow/tfjs-node-gpu');
const Jimp = require('jimp');
const path = require('path');
const fs = require('fs');
const cv = require('opencv4nodejs');

let classIndexMapping = {};
let nextClassIndex = 0;

const numClasses = () => Object.keys(classIndexMapping).length; // Get the number of classes dynamically

// Create model function
function createModel() {
  const model = tf.sequential();

  model.add(tf.layers.conv2d({
    inputShape: [50, 150, 1], // Image size
    filters: 64,
    kernelSize: 3,
    activation: 'relu',
  }));
  model.add(tf.layers.batchNormalization());
  model.add(tf.layers.maxPooling2d({ poolSize: [2, 2] }));

  model.add(tf.layers.conv2d({
    filters: 128,
    kernelSize: 3,
    activation: 'relu',
  }));
  model.add(tf.layers.batchNormalization());
  model.add(tf.layers.maxPooling2d({ poolSize: [2, 2] }));

  model.add(tf.layers.flatten());
  model.add(tf.layers.dense({ units: 512, activation: 'relu' }));
  model.add(tf.layers.dropout(0.5));

  model.add(tf.layers.dense({ units: 36 * 5, activation: 'softmax' }));

  model.compile({
    optimizer: 'adam',
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy'],
  });

  return model;
}

async function processImage(imagePath) {
  const image = await Jimp.read(imagePath);
  image.resize(150, 50).greyscale();

  const imageData = new Float32Array(image.bitmap.width * image.bitmap.height);
  for (let y = 0; y < image.bitmap.height; y++) {
    for (let x = 0; x < image.bitmap.width; x++) {
      const idx = (y * image.bitmap.width + x) * 4; // RGBA
      imageData[y * image.bitmap.width + x] = image.bitmap.data[idx] / 255.0; // Normalize
    }
  }

  return tf.tensor4d(imageData, [1, 50, 150, 1], 'float32');
}

function encodeLabel(label) {
  const charToIndex = (char) => {
    if (char >= 'a' && char <= 'z') {
      return char.charCodeAt(0) - 97; // 'a' -> 0, 'z' -> 25
    } else if (char >= '0' && char <= '9') {
      return char.charCodeAt(0) - 48 + 26; // '0' -> 26, '9' -> 35
    }
    throw new Error('Invalid character in CAPTCHA solution');
  };

  const labelTensor = label.split('').map(charToIndex);

  return tf.oneHot(tf.tensor1d(labelTensor, 'int32'), 36).reshape([5 * 36]);
}

// Prepare data: load images and labels, process them
async function prepareData(imagesDir, labels) {
    const imageFiles = fs.readdirSync(imagesDir);
    const imagePaths = imageFiles.filter(file => file.endsWith('.png'));
    
    const images = [];
    const labelIndices = []; // To store the integer class indices
  
    imagePaths.forEach(imageFile => {
      const imagePath = path.join(imagesDir, imageFile);
      const image = cv.imread(imagePath); // Load the image using OpenCV
      const resizedImage = cv.resize(image, new cv.Size(150, 50)); // Resize image to 150x50
      const normalizedImage = resizedImage.getDataAsArray(); // Normalize pixel values if necessary
  
      images.push(normalizedImage);
  
      // Get the label from the filename
      const label = labels[imageFile]; // e.g., 'ne3rr' for 'captcha4.png'
  
      // Check if label already exists in mapping, otherwise add it dynamically
      if (!(label in classIndexMapping)) {
        classIndexMapping[label] = nextClassIndex++;
      }
      
      // Push the index corresponding to the label
      labelIndices.push(classIndexMapping[label]);
    });
  
    // Convert the images array to a tensor (shape: [batch_size, height, width, channels])
    const xs = tf.tensor4d(images); 
  
    // Convert the label indices to a one-hot encoded tensor
    const ys = tf.oneHot(tf.tensor1d(labelIndices, 'int32'), numClasses()).toFloat(); // One-hot encode the labels
  
    console.log('xs shape:', xs.shape); // Should be [batch_size, height, width, channels]
    console.log('ys shape:', ys.shape); // Should be [batch_size, numClasses]
  
    return { xs, ys };
}
   
module.exports = {
  createModel,
  prepareData,
};

File: modelonly.py
import tensorflow as tf
import numpy as np
import cv2
import os

class IndexMapper:
    def __init__(self):
        self.class_index_mapping = {}
        self.next_class_index = 0

    def get_class_index(self, label):
        if label not in self.class_index_mapping:
            self.class_index_mapping[label] = self.next_class_index
            self.next_class_index += 1
        return self.class_index_mapping[label]

    def num_classes(self):
        return len(self.class_index_mapping)

# Create model function
def create_model():
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(64, (3, 3), activation='relu', input_shape=(50, 150, 1)),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu'),
        tf.keras.layers.Dropout(0.5),
        tf.keras.layers.Dense(36 * 5, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

def process_image(image_path):
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    image = cv2.resize(image, (150, 50))
    image = image.astype(np.float32) / 255.0  # Normalize image
    image = np.expand_dims(image, axis=-1)  # Add channel dimension
    return np.expand_dims(image, axis=0)  # Add batch dimension

def encode_label(label):
    char_to_index = lambda char: ord(char) - ord('a') if 'a' <= char <= 'z' else ord(char) - ord('0') + 26
    label_tensor = [char_to_index(char) for char in label]
    return tf.one_hot(label_tensor, 36)

def prepare_data(images_dir, labels):
    image_files = [f for f in os.listdir(images_dir) if f.endswith('.png')]
    images = []
    label_indices = []

    index_mapper = IndexMapper()

    for image_file in image_files:
        image_path = os.path.join(images_dir, image_file)
        image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
        image = cv2.resize(image, (150, 50))
        image = image.astype(np.float32) / 255.0  # Normalize image

        images.append(image)

        # Get the label from the filename
        label = labels.get(image_file, '')

        label_index = index_mapper.get_class_index(label)
        label_indices.append(label_index)

    images = np.array(images)
    images = np.expand_dims(images, axis=-1)  # Add channel dimension

    label_indices = np.array(label_indices)
    ys = tf.one_hot(label_indices, index_mapper.num_classes())

    print(f"xs shape: {images.shape}")
    print(f"ys shape: {ys.shape}")

    return images, ys

# Example of using the functions
if __name__ == "__main__":
    labels = {
        'captcha1.png': 'ne3rr',
        'captcha2.png': '2gjqk',
        # Add other label mappings here
    }

    images_dir = 'path_to_images'  # Specify the path to your images
    images, ys = prepare_data(images_dir, labels)

    model = create_model()
    model.fit(images, ys, epochs=10, batch_size=32)
